Libplanet/Net/Protocols/KademliaProtocol.cs

```diff
@@ -294,65 +297,105 @@ public string Trace()
             int depth,
-            Address searchTarget,
             TimeSpan? timeout,
             CancellationToken cancellationToken)
         {
             _logger.Debug(
-                $"{nameof(FindSpecificPeerAsync)}() with {{Target}} to {{Peer}}. " +
+                $"{nameof(FindSpecificPeerAsync)}() with {{Target}}. " +
                 "(depth: {Depth})",
                 target,
-                viaPeer,
                 depth);
 
-            if (history is null)
+            if (_routing.ContainsAddress(target) is BoundPeer boundPeer)
             {
-                history = new ConcurrentBag<BoundPeer>();
+                try
+                {
+                    await PingAsync(boundPeer, _requestTimeout, cancellationToken);
+                }
+                catch (PingTimeoutException)
+                {
+                    var msg =
+                        "{BoundPeer}, a target peer, is in the routing table does not respond.";
+                    _logger.Debug(msg, boundPeer);
+                    return null;
+                }
+
+                _logger.Debug(
+                    "{BoundPeer}, a target peer, is in the routing table.",
+                    boundPeer);
+                return boundPeer;
             }
 
-            IEnumerable<BoundPeer> found;
-            if (viaPeer is null)
+            var history = new ConcurrentBag<BoundPeer>();
+            var peersToFind = new ConcurrentQueue<BoundPeer>();
+            foreach (var peer in _routing.Neighbors(target, _findConcurrency, false))
             {
-                found = await QueryNeighborsAsync(history, target, timeout, cancellationToken);
+                peersToFind.Enqueue(peer);
             }
-            else
+
+            while (peersToFind.Any())
             {
-                found = await GetNeighbors(viaPeer, target, timeout, cancellationToken);
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    throw new TaskCanceledException(
+                        $"Task is cancelled during {nameof(FindSpecificPeerAsync)}()");
+                }
+
+                if (!peersToFind.TryDequeue(out BoundPeer viaPeer))
+                {
+                    continue;
+                }
+
                 history.Add(viaPeer);
-            }
+                List<BoundPeer> foundPeers =
+                    (await GetNeighbors(viaPeer, target, timeout, cancellationToken)).ToList();
 
-            BoundPeer specificPeerFound = await ProcessFoundForSpecificAsync(
-                history,
-                found,
-                target,
-                depth,
-                timeout,
-                cancellationToken,
-                searchTarget
-            );
+                int count = 0;
+                foreach (var found in foundPeers
+                    .Where(peer => !history.Contains(peer) && !peer.Address.Equals(_address))
+                    .Take(_findConcurrency))
```

Comment:
is it okay we take only this much?

Reply:
Kademlia stores closer peers in more detail. Since `foundPeers` are sorted by the distance between `Target`, this expects to only select close peers to the `Target` to reduce network load.

But as you worried, in some situation the expectation may be incorrect. In that case, we have to increase `_findConcurrency` for more accurate result.